responsemanager/server.go

```diff
@@ -229,31 +244,59 @@ func (rm *ResponseManager) processRequests(p peer.ID, requests []gsmsg.GraphSync
 			networkErrorListeners: rm.networkErrorListeners,
 			connManager:           rm.connManager,
 		}
+
+		linkSystem := result.CustomLinkSystem
+		if linkSystem.StorageReadOpener == nil {
+			linkSystem = rm.linkSystem
+		}
+
+		signals := queryexecutor.ResponseSignals{
+			PauseSignal:  make(chan struct{}, 1),
+			UpdateSignal: make(chan struct{}, 1),
+			ErrSignal:    make(chan error, 1),
+		}
+
+		responseStream := rm.responseAssembler.NewStream(rctx, p, request.ID(), subscriber)
+
+		response := &inProgressResponseStatus{
+			ctx:            rctx,
+			span:           responseSpan,
+			cancelFn:       cancelFn,
+			peer:           p,
+			request:        request,
+			linkSystem:     linkSystem,
+			customChooser:  result.CustomChooser,
+			signals:        signals,
+			startTime:      time.Now(),
+			responseStream: responseStream,
+		}
+
+		// setup query for processing
+		err := prepareQuery(rctx, p, request, result, responseStream)
+
+		if err != nil {
+			// error occurred in request hooks or setting up the query --
+			// now we're just waiting for the error response to finish sending
+			response.state = graphsync.CompletingSend
+			response.span.RecordError(err)
+			response.span.SetStatus(codes.Error, err.Error())
+		} else if result.IsPaused {
+			response.state = graphsync.Paused
+		} else {
+			response.state = graphsync.Queued
+			// TODO: Use a better work estimation metric.
+			rm.responseQueue.PushTask(p, peertask.Task{Topic: request.ID(), Priority: int(request.Priority()), Work: 1})
+		}
```

Comment:
Clarifying my understanding: If I'm reading it right, then this is the most important section of this change - prepare the query up front, and only queue the task if we've got successful setup; we also defer the traversal setup until the task gets picked up too, so we defer even more setup by doing that.

Reply:
Yes. I'm gonna comment this and rearrange for clarity